edgelet/iotedge/src/config/apply.rs

```diff
@@ -0,0 +1,440 @@
+                .ok_or_else(|| ("could not query current user information"))?;
+            (
+                current_user.clone(),
+                current_user.clone(),
+                current_user.clone(),
+                current_user.clone(),
+                current_user,
+            )
+        } else {
+            return Err("this command must be run as root".into());
+        };
+
+    let RunOutput {
+        keyd_config,
+        certd_config,
+        identityd_config,
+        tpmd_config,
+        edged_config,
+        preloaded_device_id_pk_bytes,
+    } = execute_inner(config, aziotcs_user.uid, aziotid_user.uid, iotedge_user.uid)?;
+
+    if let Some(preloaded_device_id_pk_bytes) = preloaded_device_id_pk_bytes {
+        println!("Note: Symmetric key will be written to /var/secrets/aziot/keyd/device-id");
+
+        common_config::create_dir_all("/var/secrets/aziot/keyd", &aziotks_user, 0o0700)
+            .map_err(|err| format!("{:?}", err))?;
+        common_config::write_file(
+            "/var/secrets/aziot/keyd/device-id",
+            &preloaded_device_id_pk_bytes,
+            &aziotks_user,
+            0o0600,
+        )
+        .map_err(|err| format!("{:?}", err))?;
+    }
+
+    common_config::write_file(
+        "/etc/aziot/keyd/config.d/00-super.toml",
+        &keyd_config,
+        &aziotks_user,
+        0o0600,
+    )
+    .map_err(|err| format!("{:?}", err))?;
+
+    common_config::write_file(
+        "/etc/aziot/certd/config.d/00-super.toml",
+        &certd_config,
+        &aziotcs_user,
+        0o0600,
+    )
+    .map_err(|err| format!("{:?}", err))?;
+
+    common_config::write_file(
+        "/etc/aziot/identityd/config.d/00-super.toml",
+        &identityd_config,
+        &aziotid_user,
+        0o0600,
+    )
+    .map_err(|err| format!("{:?}", err))?;
+
+    common_config::write_file(
+        "/etc/aziot/tpmd/config.d/00-super.toml",
+        &tpmd_config,
+        &aziottpm_user,
+        0o0600,
+    )
+    .map_err(|err| format!("{:?}", err))?;
+
+    common_config::write_file(
+        "/etc/aziot/edged/config.d/00-super.toml",
+        &edged_config,
+        &iotedge_user,
+        0o0600,
+    )
+    .map_err(|err| format!("{:?}", err))?;
+
+    println!("Azure IoT Edge has been configured successfully!");
+
+    // TODO: `iotedge system restart`
+
+    Ok(())
+}
+
+#[derive(Debug)]
+struct RunOutput {
+    certd_config: Vec<u8>,
+    identityd_config: Vec<u8>,
+    keyd_config: Vec<u8>,
+    tpmd_config: Vec<u8>,
+    edged_config: Vec<u8>,
+    preloaded_device_id_pk_bytes: Option<Vec<u8>>,
+}
+
+fn execute_inner(
+    config: &std::path::Path,
+    aziotcs_uid: nix::unistd::Uid,
+    aziotid_uid: nix::unistd::Uid,
+    iotedge_uid: nix::unistd::Uid,
+) -> Result<RunOutput, std::borrow::Cow<'static, str>> {
+    let config =
+        std::fs::read(config).map_err(|err| format!("could not read config file: {}", err))?;
+
+    let super_config::Config {
+        parent_hostname,
+
```

Comment:
that's some "interesting" spacing you got there (and in some other places) 👀

I get that it lines-up with how the template config "spaces" the fields, but since they'll all end up squished together once output, it doesn't really matter here. 

I don't have a super strong opinion about this, but I'd just crunch 'em down and remove the extra whitespace.

Reply:
Crunched.